### ISSUE TITLE
Minimize requirements on 3rd party libraries

### ISSUE THREAD

**Issue by omniproc** [vmware/vsphere-automation-sdk-python#92]
Hello, 

could you please consider minimizing 3rd party lib requirements? We'd really love to use the SDK rather then PyVmomi directly but currently the huge amount of 3rd party libs makes this project highly dependend on many moving parts that often break. 
Currently I see at least those requirements: 

`
certifi, urllib3, chardet, idna, requests, six, pyVmomi, suds-jurko, tabulate, setuptools, pyasn1, 
pyasn1-modules, pycparser, cffi, asn1crypto, cryptography, pyOpenSSL, attrs, service-identity, lxml, 
zope.interface, constantly, incremental, Automat, hyperlink, PyHamcrest, twisted, simplejson, werkzeug, 
vapi-runtime, vapi-client-bindings, vapi-vmc-client, vapi-common-client, nsx-policy-python-sdk
`

All of that just to deploy a VM against the SOAP API?

Also I'm not sure what's the Python 2 compatibility all about. With 6.5 we got full Python3 support in ESXi / vCenter and since Python 2 will be unsupportet with 2020, so as will be vSphere 6.0, I would imagine it will be hell alot of work to remove all the by then finally and all for good dead Python 2 workarounds. Getting rid of the `six` package would make PyVmomi way less dependend as well.

Just to give you an example: currently just in order to install the SDK on Windows one will require the `twisted` library (required by `suds-jurko` I believe). And installing twisted will fail if you don't have Visual Studio installed, since it needs some propriatary binaries (!) like `cl.exe` just to get installed.
I'm not sure how this does match up with the "open source" definition.

I mean I'm not saying everything should be done without using any 3rd party libs. But 29 3rd party libs just to make some SOAP / REST calls?

**Comment by tianhao64** [comment]
@m451 Regarding to python2, it's true that VC 6.5 has moved to python3. However, we still need to support client side code in python2. We probably can remove six and move to python3 only once python2 is deprecated. 
Do you mind sharing your use case? What are VM related tasks you are doing using Automation SDK(REST APIs)? What are the operations you can't do using Automation SDK and still need to reply on pyvmomi?

**Comment by m451** [comment]
Well in particular I was unable to get the SDK running in first place because of the said twisted requirement on .Net. To be fair I haven't spend much time looking into this issue since seeing the amount on requirements scared me away. 

But since you asked what's holding us back I'll share my thoughts:
The biggest problem I see (looking through the example and source) is that the SDK is trying to abstract the SOAP API model away, which probably is a good thing. However: as of today most DEVs working with the vSphere API atleast have some persistence across the different languages / projects (powershell, perl, python, js) / (powercli / vrealize orchestrator) because it's all the same SOAP model. 
While I think it's a good thing to finally add a layer of abstraction on that rather low-level API that PyVomi is using as well I think that that layer of sugger should be consistent with the model redesign that's currently going on with the REST API effords so that the end result is a abstracted API that is persistent across different SDKs / languages (as the SOAP model has been).
This in particular is why we still went to PyVmomi as a first step: it's a known model with little to no dependencies on 3rd party libs and very little to break. Yet: because of the low-level API access very much control. 

What's also cumber is the fact how this repo is structured. Instead of the wheel packages I'd expect to find the source in this repo and, in the releases section the *.whl files of a certain release. In addition, as another issue mentions, it would help alot if **all** the SDK packages and their dependencies would be available on PyPi. This would give development efforts the needed kickstart to start using the SDK rather then PyVmomi. 

P.S.: please don't get me wrong. I highly appreciate the efforts you're putting into this SDK and besides the JS SDK I believe that a Python SDK is probably the most relevant one. Also hats off on the samples and in-code documentation of this project, they're really great (thought Python3 typing would make the in-code documentation even better). 
But I feel like some cleanup and making the initial development easy is what's missing to get more people using the SDK rather then PyVmomi (which has the benefit of having millions of examples in blogs just because of the fact how long it's been around).

**Comment by tianhao64** [comment]
@m451 Really appreciate your feedback! I will forward your message to the team.
I'd like to quickly answer a couple of points:
1. We are working on removing twisted and service identity from the dependencies. The change should be available in upcoming releases.
2. The new REST APIs a complete new set of APIs. The new APIs are redesigned to make it much easier to use than the legacy SOAP APIs. Since many new features only support REST APIs, such as tagging, content library, HVC and etc, plus the different design and implementations between these two APIs, the automation sdk doesn't provide an abstraction layer on top of the two APIs. The Automation SDK also contains bindings based on the new REST API specs. It also ships runtime libraries to handle serialization/deserialization. 
3. The binding and runtime libraries are not open sourced yet at this point. However, we plan to upload the whls to pypi to simplify the installation process. 
4. Can you be more specific about the python3 typing suggestion?

**Comment by m451** [comment]
@tianhao64 
Nice to hear you're working on removing dependencies and pypi integration. 
I was talking about adding type hints (Python3): https://docs.python.org/3/library/typing.html#module-typing

**Comment by tianhao64** [comment]
@m451 As part of the latest runtime in [6.7u1 release](https://github.com/vmware/vsphere-automation-sdk-python/commit/1af08d664e26ade00d39edc2f6ae17cd330c20c5) we removed twisted and service-identity from the dependency list. Now 3rd party dependencies are down to:
asn1crypto, certifi, cffi, chardet, cryptography, idna, lxml, pycparser, pyOpenSSL, pyvmomi, requests, six, suds-jurko, tabulate, urllib3

We will continue to shorten this list in the future.

**Comment by tianhao64** [comment]
tabulate is removed.

**Comment by anusha94** [comment]
With the latest v7.0.1.0 SDK, the following are the list of packages installed -
`certifi, cffi, chardet, cryptography, idna, lxml, pyOpenSSL, pyVmomi, pycparser, requests, six, suds-jurko, urllib3`

I am using `Python 3.8.4`. I am not aware of any ongoing effort to reduce the dependency list further. As there is quite a decrease in number of deps as compared to when the issue was first reported, I am closing this issue. Feel free to reopen.
